data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisRecordProcessor.java

```diff
@@ -0,0 +1,173 @@
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSet;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.codec.InputCodec;
+import org.opensearch.dataprepper.model.configuration.PluginModel;
+import org.opensearch.dataprepper.model.configuration.PluginSetting;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.plugin.PluginFactory;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.kinesis.source.KinesisSource;
+import org.opensearch.dataprepper.plugins.kinesis.source.configuration.KinesisSourceConfig;
+import org.opensearch.dataprepper.plugins.kinesis.source.configuration.KinesisStreamConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.kinesis.common.StreamIdentifier;
+import software.amazon.kinesis.exceptions.InvalidStateException;
+import software.amazon.kinesis.exceptions.ShutdownException;
+import software.amazon.kinesis.exceptions.ThrottlingException;
+import software.amazon.kinesis.lifecycle.events.InitializationInput;
+import software.amazon.kinesis.lifecycle.events.LeaseLostInput;
+import software.amazon.kinesis.lifecycle.events.ProcessRecordsInput;
+import software.amazon.kinesis.lifecycle.events.ShardEndedInput;
+import software.amazon.kinesis.lifecycle.events.ShutdownRequestedInput;
+import software.amazon.kinesis.processor.RecordProcessorCheckpointer;
+import software.amazon.kinesis.processor.ShardRecordProcessor;
+import software.amazon.kinesis.retrieval.KinesisClientRecord;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+public class KinesisRecordProcessor implements ShardRecordProcessor {
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisSource.class);
+    private final StreamIdentifier streamIdentifier;
+    private final KinesisStreamConfig kinesisStreamConfig;
+    private final Duration checkpointInterval;
+    private final KinesisSourceConfig kinesisSourceConfig;
+    private final Buffer<Record<Event>> buffer;
+    private String kinesisShardId;
+    private final InputCodec codec;
+    private long lastCheckpointTimeInMillis;
+    private final int bufferTimeoutMillis;
+    private final AcknowledgementSetManager acknowledgementSetManager;
+    private final Counter acknowledgementSetCallbackCounter;
+    private final Counter recordProcessingErrors;
+    private final Counter checkpointFailures;
+    private static final Duration ACKNOWLEDGEMENT_SET_TIMEOUT = Duration.ofSeconds(20);
+    public static final String ACKNOWLEDGEMENT_SET_CALLBACK_METRIC_NAME = "acknowledgementSetCallbackCounter";
```

Comment:
I'm not sure that just getting the callbacks is a useful metric. But, having one for success and one for failure may make sense. Also, we can drop the `Counter` suffix because Micrometer will add `.count` to the name.

```
acknowledgementSetSuccesses
acknowledgementSetFailures
```